File: go/pkg/parser/nominal_type_visitor.go
package parser

import "github.com/mojo-lang/lang/go/pkg/lang"

type NominalTypeVisitor struct {
	*BaseMojoParserVisitor
}

func NewNominalTypeVisitor() *NominalTypeVisitor {
	visitor := &NominalTypeVisitor{}
	return visitor
}

func GetTypeAnnotation(ctx ITypeAnnotationContext) *lang.NominalType {
	if ctx != nil {
		visitor := NewNominalTypeVisitor()
		return ctx.Accept(visitor).(*lang.NominalType)
	}
	return nil
}

func GetType(ctx IType_Context) *lang.NominalType {
	if ctx != nil {
		visitor := NewNominalTypeVisitor()
		return ctx.Accept(visitor).(*lang.NominalType)
	}
	return nil
}

func (n *NominalTypeVisitor) VisitTypeAnnotation(ctx *TypeAnnotationContext) interface{} {
	var nominalType *lang.NominalType

	typeCtx := ctx.Type_()
	if typeCtx != nil {
		nominalType = typeCtx.Accept(n).(*lang.NominalType)
	}

	attributesCtx := ctx.Attributes()
	if attributesCtx != nil && nominalType != nil {
		visitor := NewAttributesVisitor()
		nominalType.Attributes = attributesCtx.Accept(visitor).([]*lang.Attribute)
	}

	return nominalType
}

func (n *NominalTypeVisitor) VisitType_(ctx *Type_Context) interface{} {
	basicCtx := ctx.BasicType()
	if basicCtx != nil {
		return basicCtx.Accept(n)
	}

	functionCtx := ctx.FunctionType()
	if functionCtx != nil {
		return functionCtx.Accept(n)
	}

	return nil
}

func (n *NominalTypeVisitor) VisitArrayType(ctx *ArrayTypeContext) interface{} {
	var nominalType *lang.NominalType

	typeCtx := ctx.Type_()
	if typeCtx != nil {
		nominalType = typeCtx.Accept(n).(*lang.NominalType)
	}

	attributesCtx := ctx.Attributes()
	if attributesCtx != nil {
		visitor := NewAttributesVisitor()
		nominalType.Attributes = attributesCtx.Accept(visitor).([]*lang.Attribute)
	}

	return &lang.NominalType{
		Name: "Array",
		GenericArguments: []*lang.GenericArgument{
			{
				Name:             nominalType.Name,
				GenericArguments: nominalType.GenericArguments,
				Attributes:       nominalType.Attributes,
			},
		},
	}
}

func (n *NominalTypeVisitor) VisitDictionaryType(ctx *DictionaryTypeContext) interface{} {
	keyTypeCtx := ctx.Type_(0)
	if keyTypeCtx != nil {
		keyType := keyTypeCtx.Accept(n).(*lang.NominalType)

		valueTypeCtx := ctx.Type_(1)
		if valueTypeCtx != nil {
			valueType := valueTypeCtx.Accept(n).(*lang.NominalType)

			return &lang.NominalType{
				Name: "Dictionary",
				GenericArguments: []*lang.GenericArgument{
					{
						Name:             keyType.Name,
						GenericArguments: keyType.GenericArguments,
						Attributes:       keyType.Attributes,
					},
					{
						Name:             valueType.Name,
						GenericArguments: valueType.GenericArguments,
						Attributes:       valueType.Attributes,
					},
				},
			}
		}
	}

	return nil
}

func (n *NominalTypeVisitor) VisitFunctionType(ctx *FunctionTypeContext) interface{} {
	return nil
}

func (n *NominalTypeVisitor) VisitTupleType(ctx *TupleTypeContext) interface{}  {
	return nil
}

func (n *NominalTypeVisitor) VisitTupleTypeElements(ctx *TupleTypeElementsContext) interface{} {
	return nil
}

func (n *NominalTypeVisitor) VisitUnion(ctx *UnionContext) interface{}  {
	return nil
}

func (n *NominalTypeVisitor) VisitIntersection(ctx *IntersectionContext) interface{}  {
	return nil
}

func (n *NominalTypeVisitor) VisitPrime(ctx *PrimeContext) interface{}  {
	primeCtx := ctx.PrimeType()
	if primeCtx != nil {
		return primeCtx.Accept(n)
	}

	return nil
}

func (n *NominalTypeVisitor) VisitPrimeType (ctx *PrimeTypeContext) interface{} {
	identifierCtx := ctx.TypeIdentifier()
	if identifierCtx != nil {
		visitor := NewTypeIdentifierVisitor()
		return identifierCtx.Accept(visitor)
	}

	arrayTypeCtx := ctx.ArrayType()
	if arrayTypeCtx != nil {
		return arrayTypeCtx.Accept(n)
	}

	dictTypeCtx := ctx.DictionaryType()
	if dictTypeCtx != nil {
		return dictTypeCtx.Accept(n)
	}

	return nil
}

File: java/mojo-parser/src/main/java/org/mojolang/mojo/parser/NominalTypeVisitor.java
package org.mojolang.mojo.parser;

import com.google.protobuf.GeneratedMessageV3;
import org.mojolang.mojo.lang.GenericArgument;
import org.mojolang.mojo.lang.NominalType;

public class NominalTypeVisitor extends MojoBaseVisitor<com.google.protobuf.GeneratedMessageV3> {
    private NominalType.Builder builder_ = NominalType.newBuilder();

    @Override
    public GeneratedMessageV3 visitType_annotation(MojoParser.Type_annotationContext ctx) {
        visitType_(ctx.type_());

        if (ctx.attributes() != null) {
            builder_.addAllAttributes(Util.visitAttributes(ctx.attributes()));
        }

        return builder_.build();
    }

    public GeneratedMessageV3 visitTypeAndAttributes(MojoParser.Type_Context typeCtx,
                                                     MojoParser.AttributesContext attributesCtx) {
        visitType_(typeCtx);
        if (attributesCtx != null) {
            builder_.addAllAttributes(Util.visitAttributes(attributesCtx));
        }
        return builder_.build();
    }

    @Override
    public GeneratedMessageV3 visitType_(MojoParser.Type_Context ctx) {
        MojoParser.Array_typeContext arrayCtx = ctx.array_type();
        if (arrayCtx != null) {
            NominalTypeVisitor visitor = new NominalTypeVisitor();
            NominalType nominaltype = (NominalType) visitor.visitTypeAndAttributes(arrayCtx.type_(), arrayCtx.attributes());

            GenericArgument.Builder argumentBuilder = GenericArgument.newBuilder();
            argumentBuilder.addAllAttributes(nominaltype.getAttributesList());
            argumentBuilder.setName(nominaltype.getName());

            builder_.setName("Array");
            builder_.addGenericArguments(argumentBuilder);
        }

        MojoParser.Map_typeContext mapCtx = ctx.map_type();
        if (mapCtx != null) {
        }

        MojoParser.Function_typeContext functionCtx = ctx.function_type();


        MojoParser.Type_identifierContext identifierCtx = ctx.type_identifier();
        if (identifierCtx != null) {
            TypeIdentifierVisitor visitor = new TypeIdentifierVisitor();
            NominalType type = (NominalType) visitor.visitType_identifier(identifierCtx);
            builder_.mergeFrom(type);
        }

        MojoParser.Type_Context typeCtx = ctx.type_();
        if (typeCtx != null && ctx.QUESTION() != null) {
        }

        return super.visitType_(ctx);
    }
}
